### PR TITLE
Refactor/kubernetes configs split

### LINKED ISSUE #519: Support for Kubernetes audit logs json format
The Kubernetes audit logs support two formats:

```
--audit-log-format string     Default: "json"
  | Format of saved audits. "legacy" indicates 1-line text format for each event. "json" indicates structured json format. Known formats are legacy,json.
```

The default format these days is json. However the parsing defined in kubernetes.conf is for legacy format. 
Could you please add or replace the current parsing with a json one like below ?

```
<source>
  @type tail
  @id in_tail_kube_apiserver_audit
  multiline_flush_interval 5s
  path /var/log/kubernetes/kube-apiserver-audit.log
  pos_file /var/log/kube-apiserver-audit.log.pos
  tag kube-apiserver-audit
  <parse>
    @type json
    keep_time_key true
    time_key timestamp
    time_format %Y-%m-%dT%T.%L%Z
  </parse>
</source>
```

### LINKED ISSUE #531: Split out kubernetes config file
Hi!

For ease of maintenance and customization, it would be great to have multiple config files, one for each type of kubernetes log. In fact, in my situation, I have to override the parser for audit log (see #519), but still rely on default config for the others.

Would you welcome such refactor? 
Moreover, this could be a non-breaking change I guess, because this would be feasible to change the main `kubernetes.conf` file as an includer of each sub-file. So, if it is replaced, new includes would be omitted.

In a longer sight, this would allow more souple configuration from env. Something like this in `kubernetes.conf`

```erb
include "kubernetes-audit-#{ENV['FLUENT_KUBERNETES_AUDIT_MODE'] ||"legacy"}.conf
```

### PR DESCRIPTION
**Pull request by GerkinDev**
This PR starts with a refactor of the makefile to mutualize redundant parts, to reduce the overall makefile size and increase its maintainability.

Then, each kubernetes source is splitted in its own erb.

Closes #531 
Facilitate resolution of #519 

Feedbacks welcome. More work related to #519  coming next. (I already have the fluentd conf somewhere)

### CODE REVIEW

**Comment by repeatedly** [comment]
Splitting configurations seems good.

One suggest is using `cp` is better for `conf/kubernetes/*.conf` because these configurations don't need `erb` execution.

**Comment by GerkinDev** [comment]
This is sad that this PR has been stalled and closed because of a **suggestion**. Templating is being used for many files. Static ones may be then templated without makefile change.

**Comment by cosmo0920** [comment]
Hi, this suggested changes are lack of COPY steps on Dockerfile like as:
https://github.com/fluent/fluentd-kubernetes-daemonset/blob/master/docker-image/v1.12/debian-elasticsearch7/Dockerfile#L34-L38

We should process copying files into each of images.
Without this COPY steps, users are struggled why configuration files are missing in the built containers.

**Comment by GerkinDev** [comment]
@cosmo0920 you're right.

So I'm willing to take back that work for a couple of hours in the current week to polish it. Will you be able to merge it before it goes stale ? Will I have to open another PR ?

**Comment by ashie** [comment]
I've reopened this pull request.
@GerkinDev Could you resolve conflict and resolve the thing pointed out by @cosmo0920 ?

**Comment by GerkinDev** [comment]
@ashie @cosmo0920 here you are people ! I took the freedom to rebuild al src to see what outputs looks like.

Waiting for your reviews

**Comment by GerkinDev** [comment]
Hmm, seeing the pollution in diffs inducted by the `make src-all`, would you like me to force push just 1 commit before to remove all the chore of rebuilt conf files ?

**Comment by ashie** [comment]
> Hmm, seeing the pollution in diffs inducted by the `make src-all`, would you like me to force push just 1 commit before to remove all the chore of rebuilt conf files ?

Don't worry, keep as is.
Although it seems no problem, we'll check regression. please wait for a while.

File: Makefile [hunk 1]
```diff
@@ -205,20 +204,14 @@ entrypoint.sh:
...
-		/fluent.conf.erb > docker-image/$(DOCKERFILE)/conf/fluent.conf
+	make container-image-template TEMPLATE_FILE=conf/fluent.conf
 
 # Generate kubernetes.conf from template.
 #
 # Usage:
 #	make kubernetes.conf [DOCKERFILE=] [VERSION=]
 
 kubernetes.conf:
-	mkdir -p docker-image/$(DOCKERFILE)/conf
```
Comment (ashie) [inline Makefile:221]: Memo: Just removing this line in this commit isn't appropriate, but it's cared in the following commit.

Comment (GerkinDev) [reply Makefile:221]: Oops, rebase stuff...

File: templates/conf/kubernetes.conf.erb [hunk 2]
```diff
@@ -20,172 +20,21 @@
...
-  </parse>
-</source>
+@include kubernetes/cluster-autoscaler.conf
+@include kubernetes/containers.conf
+@include kubernetes/docker.conf
+@include kubernetes/etcd.conf
+@include kubernetes/glbc.conf
+@include kubernetes/kube-apiserver-audit.conf
+@include kubernetes/kube-apiserver.conf
+@include kubernetes/kube-controller-manager.conf
```
Comment (ashie) [inline templates/conf/kubernetes.conf.erb:30]: Please use same order with `src`, or reorder `src`:

The order in `src`:
```
cluster-autoscaler.conf containers.conf docker.conf etcd.conf glbc.conf kube-apiserver-audit.conf kube-apiserver.conf kube-controller-manager.conf kubelet.conf kube-proxy.conf kube-scheduler.conf rescheduler.conf salt.conf startupscript.conf
```

**Review by ashie — CHANGES_REQUESTED** [review]

File: Makefile [hunk 3]
```diff
@@ -156,91 +156,122 @@ src-all: README.md
...
-	docker run --rm -i -v $(PWD)/templates/conf/kubernetes.conf.erb:/kubernetes.conf.erb:ro \
-		ruby:alpine erb -U -T 1 \
-			dockerfile='$(DOCKERFILE)' \
-			version='$(VERSION)' \
-		/kubernetes.conf.erb > docker-image/$(DOCKERFILE)/conf/kubernetes.conf
+	make container-image-template FILE=conf/kubernetes.conf
+	cp $(PWD)/templates/conf/tail_container_parse.conf docker-image/$(DOCKERFILE)/conf
+kubernetes.conf-all:
+	make container-image-template-all FILE=kubernetes.conf
 	cp $(PWD)/templates/conf/tail_container_parse.conf docker-image/$(DOCKERFILE)/conf
```
[the remark this answers is not in the document]
Comment (GerkinDev) [reply Makefile:239]: This was kind of refactored/copy-pasted from https://github.com/fluent/fluentd-kubernetes-daemonset/blob/5708d120d692c2bd239664a0e474025e19fa43ba/Makefile#L371-L377, but seeing that `src-all` target iterates over each docker-image by itself, I wonder if any `*-all` target is still relevant.

AFAIK the only artifact shared by all target OCI targets is the readme, which is a dependency of `src-all` anyway: https://github.com/fluent/fluentd-kubernetes-daemonset/blob/master/Makefile#L150. All others are generated on a container-basis

File: templates/conf/kubernetes.conf.erb (hunk 2 above)
Comment (GerkinDev) [reply templates/conf/kubernetes.conf.erb:30]: Resolved in c71294c1b6e11f2b6925ed45f1ec9f42dd058426

I choose to use alphabetical ascending order for both

File: Makefile [hunk 4]
```diff
@@ -156,91 +156,122 @@ src-all: README.md
...
+		/$(basename $(FILE)).erb > docker-image/$(DOCKERFILE)/$(FILE)
+
+# Render the given erb template for all images.
+#
+# Usage:
+#	make container-image-template [FILE=] [DOCKERFILE=] [VERSION=]
+
+container-image-template-all:
+	(set -e ; $(foreach img,$(ALL_IMAGES), \
+		make container-image-template $(FILE) \
```
[the remark this answers is not in the document]
Comment (GerkinDev) [reply Makefile:179]: As I mentioned in https://github.com/fluent/fluentd-kubernetes-daemonset/pull/534#discussion_r659585218, I wonder if `container-image-template-all` is not redundant with `src-all` iterating on each target artifact per container

[the remark this answers is not in the document]
Comment (ashie) [reply Makefile:179]: I'm not talking about whole of `container-image-template-all`, I'm talking about just only the following line:

```make
		make container-image-template $(FILE) \
```

It should be

```diff
-		make container-image-template $(FILE) \
+		make $(FILE) \
```

Did you test your code by yourself?
For example, did you run `make dockerfile-all`?

When I run your code I got the following error, while the original code is succeeded:

```
LANG=C make dockerfile-all

...

make[2]: *** [Makefile:166: container-image-template] Error 1
make[2]: Leaving directory '/home/aho/Projects/Fluentd/fluentd-kubernetes-daemonset-2'
make[1]: *** [Makefile:178: container-image-template-all] Error 2
make[1]: Leaving directory '/home/aho/Projects/Fluentd/fluentd-kubernetes-daemonset-2'
make: *** [Makefile:195: dockerfile-all] Error 2
```

Frankly speaking, your code is ***broken***.

**Review by ashie — CHANGES_REQUESTED** [review]

**Review by ashie — CHANGES_REQUESTED** [review]
Could you add `container-image-template` and `container-image-template-all` to `.PHONY` placed in the tail of Makefile?

**Comment by ashie** [comment]
DCO check is failed: https://github.com/fluent/fluentd-kubernetes-daemonset/pull/534/checks?check_run_id=2929861869
Could you add `Signed-off-by:` to the commit?

**Comment by GerkinDev** [comment]
As a reply to https://github.com/fluent/fluentd-kubernetes-daemonset/pull/534#discussion_r662827037, I didn't test each target, since https://github.com/fluent/fluentd-kubernetes-daemonset/blob/master/MAINTAINING.md mention only `src-all`, so I'm not aware of all the workflows or cases you expect to work, and what they should do. I've tested `make src-all` though, that does what I expect.

Moreover, `make dockerfile-all` is never mentioned anywhere, so I'm wondering if this target is really needed, since `dockerfile`s are rendered anyway with `make src`, ran on each container target in `make src-all`. Thus my comment earlier about redundant rules, that might even not be used at all.

So, am I supposed to keep all the rules previously existing ? Are they all necessary ? Weren't those legacy of older build workflows ?

Thank you

**Comment by GerkinDev** [comment]
Sorry for the kind of WIP state of the PR lately. I'm cleaning it up, and hope that this one will be the good one.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (GerkinDev) [reply Makefile]: Indeed, I realized that `container-image-template-all` target wasn't appropriate. I've replaced with `each-image TARGET=` target which runs the given `TARGET` for each images, allowing proper support of additional sub-targets commands such as `cp` or anything.

[the remark this answers is not in the document]
Comment (ashie) [reply Makefile]: Good naming! :+1:

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashie) [reply Makefile]: Fixed in 83878ad949a5f75c79afbf608505b43a528f908e

File: Makefile [hunk 5]
```diff
@@ -268,10 +353,9 @@ plugins:
 #
 # Usage:
 #	make post-checkout-hook [DOCKERFILE=]
-
 post-checkout-hook:
 	if [ -n "$(findstring /arm64/,$(DOCKERFILE))" ]; then \
-		mkdir -p $(DOCKERFILE)/hooks; \
+		mkdir -p docker-image/$(DOCKERFILE)/hooks; \
```
Comment (ashie) [inline Makefile:358]: Good catch!
I wondered why broken directories are created after `make src-all` for a while ago. This is the answer.

**Comment by ashie** [comment]
LGTM, thanks for you work!
I think a remaining task is adding a commit for generated files, especially for conf/kubernetes/*.conf for each images.
Otherwise we might forget to include them into images when we publish them.

@cosmo0920 Could you double-check?
Note that this PR includes a trivial bug fix: https://github.com/fluent/fluentd-kubernetes-daemonset/pull/534/files#r663615221

**Review by cosmo0920 — APPROVED** [review]
LGTM.
Split configurations and generating them work for me.

**Comment by GerkinDev** [comment]
Okay I'll `make src-all` a bit later once I'm back on my Linux env. Please wait a couple of hours

**Comment by cosmo0920** [comment]
> Okay I'll make src-all a bit later once I'm back on my Linux env. Please wait a couple of hours

Ah, you don't have to generate files with make src-all.
It will make severe conflicts if we do more time for reviewing. Merging this PR and generating files with make  src-all is executed by us is more safe for conflicts.

**Comment by GerkinDev** [comment]
So do you need anything else?

**Comment by cosmo0920** [comment]
@ashie @kenhys Can we merge this?

**Review by ashie — APPROVED** [review]

**Comment by ashie** [comment]
:+1:

**Comment by cosmo0920** [comment]
I'd applied this patch at master: https://github.com/fluent/fluentd-kubernetes-daemonset/commit/87ec024646dd3d1749caeb035be94898a90e8c99
And I kicked Fluentd kubernetes daemonset in DockerHub.

**Comment by ashie** [comment]
Hmm, `in_tail_container_logs` is duplicated: #583

**Pull request merged**